pandas/core/internals/blocks.py

```diff
@@ -993,15 +992,16 @@ def putmask(self, mask, new, align=True, inplace=False, axis=0,
             #
             # TODO: this prob needs some better checking
             # for 2D cases
-            if ((is_list_like(new) and
-                 np.any(mask[mask]) and
+            if ((is_list_like(new) and np.any(mask[mask]) and
                  getattr(new, 'ndim', 1) == 1)):
-
                 if not (mask.shape[-1] == len(new) or
                         mask[mask].shape[-1] == len(new) or
                         len(new) == 1):
-                    raise ValueError("cannot assign mismatch "
-                                     "length to masked array")
+                    # GH 19266 and GH 21977
+                    # ValueError triggers try except block in Block.replace
+                    # causing RecursionError
+                    raise Exception("cannot assign mismatch length "
```

Comment:
can you make this a bit more informative from a user POV?

Reply:
sure. this change here is exposed to user so when one tries to assign an array of n-dimension to mask of m-dimension, when m != n, this error throws. 

```python
import pandas as pd

s = pd.Series(range(10))
mask = s > 5

>>> s[mask].shape
(4,)

s[mask] = [5, 4, 3, 2, 1]

  File "/Users/mingli/.virtualenvs/cust-value/lib/python3.5/site-packages/pandas/core/internals.py", line 1006, in putmask
    raise ValueError("cannot assign mismatch "
ValueError: cannot assign mismatch length to masked array
```
This is also used internally by `replace` and ValueError it raises is causing infinite loop as per linked issues. Whilst changing that to not ValueError will solve it, it also makes it less clear and potentially breaks users' codes if they expect to catch ValueError running above code snippet.

Do you think it's ok to use Exception here as it would solve the issues linked but users might have to change their code?

Reply:
this is reverted.